_posts/2020-08-05-using-open-policy-agent-with-strimzi-and-apache-kafka.md

```diff
@@ -0,0 +1,455 @@
+
+```rego
+is_consumer_group {
+    user_groups(principal.name)[_] == topic_consumer_groups(topic_name)[_]
+}
+
+is_producer_group {
+    user_groups(principal.name)[_] == topic_producer_groups(topic_name)[_]
+}
+
+user_groups(user) = groups {
+    groups := json.unmarshal(kafkausers[_][user].metadata.annotations["groups"])
+}
+
+topic_consumer_groups(topic) = groups {
+    groups := json.unmarshal(kafkatopics[_][topic].metadata.annotations["consumer-groups"])
+}
+
+topic_producer_groups(topic) = groups {
+    groups := json.unmarshal(kafkatopics[_][topic].metadata.annotations["producer-groups"])
+}
+```
+
+The policy above will extract the groups from the custom resource annotations.
+And see if there is any group shared between the user and the topic.
+If any group is in both user groups and consumer or producer groups, the user will be allowed.
+
+The admin operations are handled differently and any user who is member of the special group `admin` will be allowed to do the admin operations:
+
+```rego
+admin_groups := ["admin"]
+
+is_admin_group {
+    user_groups(principal.name)[_] == admin_groups[_]
+}
+```
+
+Now we just need to use these rules to make the `allow` or `deny` decision:
+
+```rego
+allow {
+    not deny
+}
+
+deny {
+    not allow_consumer_topic
+    not allow_consumer_group
+    not allow_producer
+    not allow_admin
+}
+
+allow_consumer_topic {
+    is_topic_resource
+    is_consumer_operation
+    is_consumer_group
+}
+
+allow_consumer_group {
+    is_group_resource
+    is_consumer_operation
+    startswith(group_name, principal.name)
+}
+
+allow_producer {
+    is_topic_resource
+    is_producer_operation
+    is_producer_group
+}
+
+allow_admin {
+    is_admin_operation
+    is_admin_group
+}
+```
+
+The full policy including some additional helper functions can be found on my [GitHub](https://github.com/scholzj/demo-opa-kafka-authorization/blob/master/advanced-example-policy.rego).
+
+## Conclusion
+
+The main advantage of Open Policy Agent authorization compared to the `simple` or Keycloak authorizers is its flexibility.
+The examples from the previous sections demonstrate how easily you can use OPA and Rego to combine data from different sources and use them for the authorization decisions.
+
+With the `simple` authorizer, you have to follow the ACL rules defined for each user and stored in ZooKeeper.
+With Keycloak, you can use the permissions defined in Keycloak.
+But with OPA, you can use any external data you want.
+OPA calls this _context-aware authorization and decisions_.
+You can [easily integrate](https://www.openpolicyagent.org/docs/latest/external-data/) OPA with different external data sources and use these data for the decision making to do exactly what you want.
+
+The main disadvantage is that unless you are already using OPA for other applications, you will need to run yet another application.
```

Comment:
```suggestion
The main disadvantage is that unless you are already using OPA for other applications, you will need to run another application.
```

I think yet another sounds like this is more negative than it may well be.